File: ra-idp-go/internal/tenancy/usecases/manage_tenants.go
package usecases

import (
	"context"
	"errors"
	"strings"
	"time"

	"ra-idp-go/internal/spec"
	tenantports "ra-idp-go/internal/tenancy/ports"
)

var (
	ErrTenantNotFound       = errors.New("tenant not found")
	ErrTenantConflict       = errors.New("tenant already exists")
	ErrInvalidTenantID      = errors.New("invalid tenant id")
	ErrDefaultTenant        = errors.New("default tenant cannot be disabled")
	ErrDisplayNameEmpty     = errors.New("display name is required")
	ErrPolicyOverrideWeaker = errors.New("password policy override is weaker than the global default")
)

// UpdateInput はテナント設定の部分更新を表す。nil のフィールドは現状維持。
// PasswordPolicyOverride にゼロ値を渡すと override を解除する (global default 継承)。
type UpdateInput struct {
	DisplayName            *string
	PasswordPolicyOverride *spec.PasswordPolicyOverride
}

// PolicyFloor は password_policy_override が下回ってはならない global 値。
// MinLength の最低値、MaxLength の上限値、HistoryDepth の最低値で gating する (WI-17)。
type PolicyFloor struct {
	MinLength    int
	MaxLength    int
	HistoryDepth int
}

func EnsureDefault(ctx context.Context, repo tenantports.TenantRepository, now time.Time) error {
	tenant, err := repo.FindByID(ctx, spec.DefaultTenantID)
	if err != nil {
		return err
	}
	if tenant != nil {
		return nil
	}
	now = normalizeNow(now)
	return repo.Save(ctx, &spec.Tenant{
		ID: spec.DefaultTenantID, DisplayName: "Default", Status: spec.TenantStatusActive, CreatedAt: now,
	})
}

func Create(ctx context.Context, repo tenantports.TenantRepository, id, displayName string, now time.Time) (*spec.Tenant, error) {
	displayName = strings.TrimSpace(displayName)
	if displayName == "" {
		return nil, ErrDisplayNameEmpty
	}
	tenant := &spec.Tenant{
		ID: strings.TrimSpace(id), DisplayName: displayName, Status: spec.TenantStatusActive,
		CreatedAt: normalizeNow(now),
	}
	if err := tenant.Validate(); err != nil {
		return nil, ErrInvalidTenantID
	}
	existing, err := repo.FindByID(ctx, tenant.ID)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		return nil, ErrTenantConflict
	}
	if err := repo.Save(ctx, tenant); err != nil {
		return nil, err
	}
	return tenant, nil
}

func Update(
	ctx context.Context,
	repo tenantports.TenantRepository,
	id string,
	input UpdateInput,
	floor PolicyFloor,
	now time.Time,
) (*spec.Tenant, error) {
	tenant, err := repo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if tenant == nil {
		return nil, ErrTenantNotFound
	}
	updated := *tenant
	if input.DisplayName != nil {
		name := strings.TrimSpace(*input.DisplayName)
		if name == "" {
			return nil, ErrDisplayNameEmpty
		}
		updated.DisplayName = name
	}
	if input.PasswordPolicyOverride != nil {
		normalized := normalizeOverride(*input.PasswordPolicyOverride)
		if normalized == nil {
			updated.PasswordPolicyOverride = nil
		} else {
			if err := enforcePolicyFloor(*normalized, floor); err != nil {
				return nil, err
			}
			updated.PasswordPolicyOverride = normalized
		}
	}
	t := normalizeNow(now)
	updated.UpdatedAt = &t
	if err := repo.Save(ctx, &updated); err != nil {
		return nil, err
	}
	return &updated, nil
}

func normalizeOverride(o spec.PasswordPolicyOverride) *spec.PasswordPolicyOverride {
	result := spec.PasswordPolicyOverride{}
	any := false
	if o.MinLength != nil && *o.MinLength > 0 {
		v := *o.MinLength
		result.MinLength = &v
		any = true
	}
	if o.MaxLength != nil && *o.MaxLength > 0 {
		v := *o.MaxLength
		result.MaxLength = &v
		any = true
	}
	if o.HistoryDepth != nil && *o.HistoryDepth > 0 {
		v := *o.HistoryDepth
		result.HistoryDepth = &v
		any = true
	}
	if !any {
		return nil
	}
	return &result
}

func enforcePolicyFloor(o spec.PasswordPolicyOverride, floor PolicyFloor) error {
	if o.MinLength != nil && floor.MinLength > 0 && *o.MinLength < floor.MinLength {
		return ErrPolicyOverrideWeaker
	}
	if o.MaxLength != nil && floor.MaxLength > 0 && *o.MaxLength > floor.MaxLength {
		return ErrPolicyOverrideWeaker
	}
	if o.HistoryDepth != nil && floor.HistoryDepth > 0 && *o.HistoryDepth < floor.HistoryDepth {
		return ErrPolicyOverrideWeaker
	}
	return nil
}

func SetDisabled(ctx context.Context, repo tenantports.TenantRepository, id string, disabled bool, now time.Time) (*spec.Tenant, error) {
	if id == spec.DefaultTenantID && disabled {
		return nil, ErrDefaultTenant
	}
	tenant, err := repo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if tenant == nil {
		return nil, ErrTenantNotFound
	}
	updated := *tenant
	t := normalizeNow(now)
	updated.UpdatedAt = &t
	if disabled {
		updated.Status = spec.TenantStatusDisabled
		updated.DisabledAt = &t
	} else {
		updated.Status = spec.TenantStatusActive
		updated.DisabledAt = nil
	}
	if err := repo.Save(ctx, &updated); err != nil {
		return nil, err
	}
	return &updated, nil
}

func normalizeNow(now time.Time) time.Time {
	if now.IsZero() {
		return time.Now().UTC()
	}
	return now.UTC()
}

File: ra-idp/src/tenancy/usecases/manage-tenants.ts
/**
 * テナント lifecycle ユースケース (ADR-032)。
 *
 * Create / Update / SetDisabled / EnsureDefault。
 * default テナントは起動時に idempotent に upsert され、disable は不可。
 */

import { DEFAULT_TENANT_ID, TENANT_ID_PATTERN, TenantSchema, type Tenant } from '../../spec-bindings/schemas'
import type { TenantRepository } from '../ports/tenant-repository'

export class TenantNotFoundError extends Error {
  constructor() {
    super('tenant not found')
    this.name = 'TenantNotFoundError'
  }
}

export class TenantConflictError extends Error {
  constructor() {
    super('tenant already exists')
    this.name = 'TenantConflictError'
  }
}

export class InvalidTenantIdError extends Error {
  constructor() {
    super('invalid tenant id')
    this.name = 'InvalidTenantIdError'
  }
}

export class DisplayNameRequiredError extends Error {
  constructor() {
    super('display name is required')
    this.name = 'DisplayNameRequiredError'
  }
}

export class DefaultTenantImmutableError extends Error {
  constructor() {
    super('default tenant cannot be disabled')
    this.name = 'DefaultTenantImmutableError'
  }
}

export async function ensureDefaultTenant(
  repo: TenantRepository,
  now: Date = new Date(),
): Promise<void> {
  const existing = await repo.findById(DEFAULT_TENANT_ID)
  if (existing) return
  await repo.save(
    TenantSchema.parse({
      id: DEFAULT_TENANT_ID,
      display_name: 'Default',
      status: 'active',
      created_at: now.toISOString(),
    }),
  )
}

export async function createTenant(
  repo: TenantRepository,
  input: { id: string; display_name: string },
  now: Date = new Date(),
): Promise<Tenant> {
  const id = input.id.trim()
  const displayName = input.display_name.trim()
  if (!displayName) throw new DisplayNameRequiredError()
  if (!TENANT_ID_PATTERN.test(id) || id === 'admin') throw new InvalidTenantIdError()

  const existing = await repo.findById(id)
  if (existing) throw new TenantConflictError()

  const tenant = TenantSchema.parse({
    id,
    display_name: displayName,
    status: 'active',
    created_at: now.toISOString(),
  })
  await repo.save(tenant)
  return tenant
}

export async function updateTenant(
  repo: TenantRepository,
  id: string,
  input: { display_name: string },
  now: Date = new Date(),
): Promise<Tenant> {
  const displayName = input.display_name.trim()
  if (!displayName) throw new DisplayNameRequiredError()
  const tenant = await repo.findById(id)
  if (!tenant) throw new TenantNotFoundError()
  const updated: Tenant = {
    ...tenant,
    display_name: displayName,
    updated_at: now.toISOString(),
  }
  await repo.save(updated)
  return updated
}

export async function setTenantDisabled(
  repo: TenantRepository,
  id: string,
  disabled: boolean,
  now: Date = new Date(),
): Promise<Tenant> {
  if (id === DEFAULT_TENANT_ID && disabled) throw new DefaultTenantImmutableError()
  const tenant = await repo.findById(id)
  if (!tenant) throw new TenantNotFoundError()
  const updated: Tenant = {
    ...tenant,
    status: disabled ? 'disabled' : 'active',
    disabled_at: disabled ? now.toISOString() : undefined,
    updated_at: now.toISOString(),
  }
  await repo.save(updated)
  return updated
}
